arch/x86/core/early_serial.c

```diff
@@ -1,102 +1,67 @@
-
-#define IER_DISABLE		0x00
-#define LCR_8N1			(BIT(0) | BIT(1))
-#define LCR_DLAB_SELECT		BIT(7)
-#define MCR_DTR			BIT(0)
-#define MCR_RTS			BIT(1)
-#define LSR_THRE		BIT(5)
-
-#define FCR_FIFO		BIT(0)	/* enable XMIT and RCVR FIFO      */
-#define FCR_RCVRCLR		BIT(1)	/* clear RCVR FIFO                */
-#define FCR_XMITCLR		BIT(2)	/* clear XMIT FIFO                */
-#define FCR_FIFO_1		0	/* 1 byte in RCVR FIFO            */
-
-/* convenience defines */
-#define REG_THR(x)		(x + REG_OFFSET_THR * UART_REG_ADDR_INTERVAL)
-#define REG_IER(x)		(x + REG_OFFSET_IER * UART_REG_ADDR_INTERVAL)
-#define REG_FCR(x)		(x + REG_OFFSET_FCR * UART_REG_ADDR_INTERVAL)
-#define REG_LCR(x)		(x + REG_OFFSET_LCR * UART_REG_ADDR_INTERVAL)
-#define REG_MCR(x)		(x + REG_OFFSET_MCR * UART_REG_ADDR_INTERVAL)
-#define REG_LSR(x)		(x + REG_OFFSET_LSR * UART_REG_ADDR_INTERVAL)
-#define REG_BRDL(x)		(x + REG_OFFSET_BRDL * UART_REG_ADDR_INTERVAL)
-#define REG_BRDH(x)		(x + REG_OFFSET_BRDH * UART_REG_ADDR_INTERVAL)
-
-#if DT_NODE_HAS_PROP(DT_CHOSEN(zephyr_console), reg_shift)
-#define UART_REG_ADDR_INTERVAL \
-	(1 << DT_PROP(DT_CHOSEN(zephyr_console), reg_shift))
-#endif
-
+/* Legacy I/O Port Access to a NS16550 UART */
 #ifdef UART_NS16550_ACCESS_IOPORT
-#define PORT		((io_port_t)DT_REG_ADDR(DT_CHOSEN(zephyr_console)))
-#define INBYTE(x)	sys_in8(x)
-#define INWORD(x)	sys_in32(x)
-#define OUTBYTE(x, d)	sys_out8(d, x)
-#define OUTWORD(x, d)	sys_out32(d, x)
-#ifndef UART_REG_ADDR_INTERVAL
-#define UART_REG_ADDR_INTERVAL 1 /* address diff of adjacent regs. */
-#endif /* UART_REG_ADDR_INTERVAL */
-#else
-#define PORT		((mm_reg_t)DT_REG_ADDR(DT_CHOSEN(zephyr_console)))
-#define INBYTE(x)	sys_read8(x)
-#define INWORD(x)	sys_read32(x)
-#define OUTBYTE(x, d)	sys_write8(d, x)
-#define OUTWORD(x, d)	sys_write32(d, x)
-#ifndef UART_REG_ADDR_INTERVAL
-#define UART_REG_ADDR_INTERVAL 4 /* address diff of adjacent regs. */
+#define IN(reg)       sys_in8(reg)
+#define OUT(reg, val) sys_out8(val, reg)
 #endif
-#endif /* UART_NS16550_ACCESS_IOPORT */
 
-#ifdef CONFIG_UART_NS16550_ACCESS_WORD_ONLY
-#undef INBYTE
-#define INBYTE(x) INWORD(x)
-#undef OUTBYTE
-#define OUTBYTE(x, d) OUTWORD(x, d)
+/* "Modern" mapping of a UART into a PCI MMIO device.  The registers
+ * are still bytes, but spaced at a 32 bit stride instead of packed
+ * together.
+ */
+#ifdef X86_SOC_EARLY_SERIAL_PCIDEV
+static uintptr_t pci_bar;
+#define IN(reg)       (sys_read32(pci_bar + reg * 4) & 0xff)
+#define OUT(reg, val) sys_write32((val) & 0xff, pci_bar + reg * 4)
 #endif
 
-#ifdef UART_NS16550_PCIE_ENABLED
-static mm_reg_t base;
-#else
-#define base PORT
+/* Still other devices use a MMIO region containing packed byte
+ * registers
+ */
+#ifdef X86_SOC_EARLY_SERIAL_MMIO8_ADDR
+#define IN(reg)       sys_read8(X86_SOC_EARLY_SERIAL_MMIO8_ADDR + reg)
+#define OUT(reg, val) sys_write8(val, X86_SOC_EARLY_SERIAL_MMIO8_ADDR + reg)
 #endif
 
+#define REG_THR  0x00  /* Transmitter holding reg. */
```

Comment:
I wonder if it makes sense to introduce a common ns16550.h with these definitions under drivers/serial and have both this file and the real 16550 driver use it.